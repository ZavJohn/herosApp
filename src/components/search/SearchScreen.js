import React, { useMemo }  from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import querytString from 'query-string';
import { getHeroByName } from '../../helpers/getHeroByName';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../hero/HeroCard';

export const SearchScreen = () => {
  
 const navigate = useNavigate();
 const location = useLocation();

const { q = ''} = querytString.parse(location.search);
 
  const [ formValues, handleInputChange, reset ] = useForm({
   searchText: q,
 });

 const { searchText } = formValues;

 const heroesFilter = useMemo( getHeroByName(q), [ q ] );
 
 const handleSearch = (e) => {
    e.preventDefault();

    navigate(`?q=${ searchText }`);

  };

  return (
    <>
        <h1>Búsquedas</h1>
        <hr />

        <div className="row">
          <div className="col-5">
            <h4>Buscar</h4>
            <hr />

            <form className="row" onSubmit={ handleSearch }>
              <div className="input-group mb-3">
                <input
                  type="text"
                  placeholder="Ingresa tu búsqueda"
                  className="form-control"
                  name="searchText"
                  autoComplete="off"
                  value={ searchText }
                  onChange={ handleInputChange }
                />

                <button
                  className="input-group-text btn-secondary"
                  type="submit"
                >
                  Buscar
                </button>
              </div>
            </form>
          </div>

          <div className="col-7">
            <h4>Resultados</h4>
            <hr />

            {
              (q==='')
                ? <div className="alert alert-info">No se ingresaron criterioz de búsqueda</div>
                    : ( heroesFilter.length === 0 ) 
                        && <div className="alert alert-danger">No se encontraron resultados en: { q }</div>
            }
            { 
              heroesFilter.map(hero =>(
                  <HeroCard 
                    key={ hero.id }
                    { ...hero }
                  />
              ))
            }

          </div>  
        </div>
    </> 
  )
}
