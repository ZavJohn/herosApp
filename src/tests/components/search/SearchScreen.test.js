import { mount } from 'enzyme';
import { SearchScreen } from '../../../components/search/SearchScreen';
import { MemoryRouter, useNavigate } from 'react-router-dom';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}));
describe('Pruebas en el <SearchScreen />', () => {
    
    test('Debe mostrarse correctamente con valores por defecto', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={[ '/search' ]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-info').text().trim() ).toBe('No se ingresaron criterios de búsqueda');
    });

    test('Debe mostrar a Batman y el input con el queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={[ '/search?q=batman' ]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper.find('input').prop('value') ).toBe('batman');
    
    });

    test('Debe mostrar mensaje de alerta por busqueda erronea', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={[ '/search?q=123' ]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper.find('.alert-danger').text().trim() ).toBe('No se encontraron resultados en: 123');
    
    });

    test('Acción del submit', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={[ '/search?q=123' ]}>
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'hulk'
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });
        
        expect( mockNavigate ).toHaveBeenCalledWith('?q=hulk');
        //expect( wrapper.find('.alert-danger').text().trim() ).toBe('No se encontraron resultados en: 123');
    
    });

})