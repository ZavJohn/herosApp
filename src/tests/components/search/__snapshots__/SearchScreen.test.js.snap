// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <SearchScreen /> Debe mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Búsquedas
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        className="row"
        onSubmit={[Function]}
      >
        <div
          className="input-group mb-3"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="searchText"
            onChange={[Function]}
            placeholder="Ingresa tu búsqueda"
            type="text"
            value=""
          />
          <button
            className="input-group-text btn-secondary"
            type="submit"
          >
            Buscar
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        No se ingresaron criterios de búsqueda
      </div>
    </div>
  </div>,
]
`;
